### ISSUE TITLE
pyright passing AnyStr to AnyStr (os.scandir) results in error

### ISSUE THREAD

**Issue by shoffmeister** [microsoft/pylance-release#1918]
Issue Type: <b>Bug</b>

Passing a variable of type `AnyStr` to `os.scandir` (which has an overload of `def scandir(path: AnyStr | PathLike[AnyStr])`) yields `No overloads for "scandir" match the provided arguments - Argument types: (AnyStr@thing)`

```python
import os
from typing import AnyStr

def thing(value: AnyStr):
    with os.scandir(value) as it:
        for file in it:
            if file.is_file(): ...

thing("Hello")   # proper string
thing(b"world")  # bytes

thing(None) # just to make a point: `None` is incompatible with AnyStr
```

This is on Python 3.9.7.

FWIW, mypy does not complain.

Extension version: 2021.10.0
VS Code version: Code 1.61.0 (ee8c7def80afc00dd6e593ef12f37756d8f504ea, 2021-10-07T18:13:09.652Z)
OS version: Windows_NT x64 10.0.18363
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz (8 x 2904)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|63.66GB (40.41GB free)|
|Process Argv|--crash-reporter-id 5ddcd13a-5853-4d88-9c59-31e3be66729a|
|Screen Reader|no|
|VM|44%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383cf:30185419
pythonvspyt602:30300191
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythonvspyt639:30300192
pythontb:30283811
pythonvspyt551cf:30345471
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscorecescf:30358481
pythondataviewer:30285071
pythonvsuse255:30340121
vscod805:30301674
pythonvspyt200:30340761
binariesv615:30325510
vsccppwtct:30378365
pythonvssor306:30344512
bridge0708:30335490
pygetstartedt2:30371810
dockerwalkthru:30377721
bridge0723:30353136
pythonrunftest32:30373476
pythonf5test824:30373475
javagetstartedc:30364665
pythonvspyt187:30373474
pydsgst2:30361792
vsqsis200:30374795
vsaa593cf:30376535

```

</details>

**Comment by shoffmeister** [comment]
FWIW, this being Python 3.9.7, I am a bit irritated to have the vscode editor type hover show me
```
(function)
scandir(path: None = ...) -> _ScandirIterator[str]
scandir(path: int) -> _ScandirIterator[str]

scandir(path: AnyStr@scandir | PathLike[AnyStr@scandir]) -> _ScandirIterator[AnyStr@scandir]
```
because, according to https://www.python.org/dev/peps/pep-0604/ the | operator would only work as of Python 3.10?

Also, the typeshed-fallback from the extension distribution features
```python
if sys.version_info >= (3, 7):
    @overload
    def scandir(path: None = ...) -> _ScandirIterator[str]: ...
    @overload
    def scandir(path: int) -> _ScandirIterator[str]: ...
    @overload
    def scandir(path: AnyStr | PathLike[AnyStr]) -> _ScandirIterator[AnyStr]: ...

else:
```
while I would expect that the last of the overloads is protected by `sys.version_info >= (3, 10)` due to the syntax chosen?

(but then I have no idea about typeshed and no deeper insight into the Python type system at the moment, sorry)

**Comment by erictraut** [comment]
Thanks for reporting the issue. I agree this is a bug. I've created a pyright tracking issue and will investigate further.

**Comment by bschnurr** [comment]
This issue has been fixed in version 2021.10.1, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2021100-6-october-2021
